Feature/hostname redirect

mobinzk opened the pull request:
Using hostname redirect module allows users to:
- Force SSL (Force visitors to use SSL (https). This is a basic implementation that will change the URL from http to https.)
- Redirect to none WWW, to WWW or a custom domain

mobinzk committed:
Add foundation for hostname redirect

Added menu item/permission

mobinzk committed:
Add hostname setting fields

Also added the save ability into driver

mobinzk committed:
Toggle redirect to site URL when redirect option select changes

mobinzk committed:
Add hostname redirects

Using hostname redirect module allows users to:
- Force SSL (Force visitors to use SSL (https). This is a basic implementation that will change the URL from http to https.)
- Redirect to none WWW, to WWW or a custom domain

peterkeating commented on HostnameRedirects/Services/RewriteOptionsService.cs:
```diff
@@ -0,0 +1,138 @@
+
+        #region Helpers
+
+        private string getURL(RewriteContext context) {
```
This should be `GetUrl`, method names in C# should always start with a capital.

peterkeating commented on HostnameRedirects/Services/RewriteOptionsService.cs:
```diff
@@ -0,0 +1,138 @@
+
+        private string getURL(RewriteContext context) {
+            var request = context.HttpContext.Request;
+            var url = request.Scheme + "://" + request.Host.Value + request.PathBase + request.Path + request.QueryString;
```
Instead of defining `url` here and then returning it on the next line, just return straight away.

Also you can improve the construction of the string using the improved way of string interpolation, shown below.

`return $"{request.Scheme}://{request.Host.Value}{request.PathBase}{request.Path}{request.QueryString}`

peterkeating commented on HostnameRedirects/Services/RewriteOptionsService.cs:
```diff
@@ -0,0 +1,138 @@
+            }
+
+
+            var ignoredUrls = settings.IgnoredUrls.Split(new string[] { "\n" }, StringSplitOptions.RemoveEmptyEntries);
```
Looks like this isn't used so let's remove it :)

peterkeating commented on HostnameRedirects/Services/RewriteOptionsService.cs:
```diff
@@ -0,0 +1,138 @@
+
+            var ignoredUrls = settings.IgnoredUrls.Split(new string[] { "\n" }, StringSplitOptions.RemoveEmptyEntries);
+
+            return settings.IgnoredUrls.Split(new string[] { "\n" }, StringSplitOptions.RemoveEmptyEntries).Any(x => url.StartsWith(x));
```
Instead of using `\n` you should use `System.Environment.NewLine`

More information about why can be found here - https://github.com/OrchardCMS/OrchardCore/wiki/Engineering-Guidelines#line-breaks

peterkeating commented on HostnameRedirects/Services/RewriteOptionsService.cs:
```diff
@@ -0,0 +1,138 @@
+        #endregion
+
+        public int StatusCode { get; } = (int) HttpStatusCode.MovedPermanently;
+        public ILogger Logger { get; set; } = new NullLogger();
```
This is a dependency, so I'd move it to the dependency region.

peterkeating commented on HostnameRedirects/Services/RewriteOptionsService.cs:
```diff
@@ -0,0 +1,138 @@
+
+        #endregion
+
+        public int StatusCode { get; } = (int) HttpStatusCode.MovedPermanently;
```
Maybe create a new region for "Properties" that'll house the `StatusCode`.

mobinzk committed:
Fix method name consistency

mobinzk committed:
Fix method name

mobinzk committed:
Improve GetURL method

mobinzk committed:
Remove unused variable

mobinzk committed:
Using Environment.NewLine instead of \n

More information about why can be found here - https://github.com/OrchardCMS/OrchardCore/wiki/Engineering-Guidelines#line-breaks

mobinzk committed:
Move ILogger dependency to dependencies region

mobinzk committed:
Improve regions

mobinzk committed:
Add hostname redirects to readme

peterkeating commented:
One last thing @mobinzk, can you add a section in the README for hostname redirects (above "Redirects" so they are in alphabetical order).

mobinzk committed:
Sort features headings alphabetically

peterkeating commented:
Ahh didn't see you had already added it to the README :) 

Code's looking good, will give it a run through when I get a moment 👍
